browser/brave_ads/ads_service_impl.cc

```diff
@@ -2098,6 +2106,96 @@ void AdsServiceImpl::RecordP2AEvent(const std::string& name,
   }
 }
 
+void AdsServiceImpl::LogTrainingCovariates(
+    const ads::mojom::TrainingInstancePtr instance) {
+  if (!ad_notification_timing_data_store_) {
+    return;
+  }
+
+  // TODO(https://github.com/brave/brave-browser/issues/21189): Refactor DB to
+  // use generic key/value schema across all data stores
+  brave_federated::AdNotificationTimingTaskLog log;
+
+  for (const auto& covariate : instance->covariates) {
+    switch (covariate->covariate_type) {
+      case ads::mojom::CovariateType::kAdNotificationWasClicked: {
+        if (covariate->data_type != ads::mojom::DataType::kBool) {
+          VLOG(0) << "covariate type not of type bool";
+          break;
+        }
+
+        bool value_as_bool = false;
+        if (covariate->value == kTrue) {
+          value_as_bool = true;
+        } else if (covariate->value != kFalse) {
+          NOTREACHED() << "Failed to convert covariate value to bool";
+          continue;
+        }
+
+        log.label = value_as_bool;
+        break;
+      }
+
+      case ads::mojom::CovariateType::kAdNotificationLocaleCountryAtTimeOfServing: {
+        if (covariate->data_type != ads::mojom::DataType::kString) {
+          VLOG(0) << "covariate type not of type string";
+          break;
+        }
+
+        log.locale = covariate->value;
+        break;
+      }
+
+      case ads::mojom::CovariateType::kAdNotificationImpressionServedAt: {
+        if (covariate->data_type != ads::mojom::DataType::kDouble) {
+          VLOG(0) << "covariate type not of type double";
+          break;
+        }
+
+        double value_as_double = 0.0;
+        if (!base::StringToDouble(covariate->value, &value_as_double)) {
+          NOTREACHED() << "Failed to convert covariate value to double";
+          continue;
+        }
+
+        log.time = base::Time::FromDoubleT(value_as_double);
+        break;
+      }
+
+      case ads::mojom::CovariateType::kAdNotificationNumberOfTabsOpenedInPast30Minutes: {
+        if (covariate->data_type != ads::mojom::DataType::kInt64) {
```

Comment:
why are we using `int64` for number of opened tabs? `int` is more suitable for this value.

Reply:
did change it to `int` based on @tmancey comment, forgot to change it everywhere.